pyro/contrib/gp/models/gpr.py

```diff
@@ -5,81 +5,94 @@
+
+        self.kernel_prior = kernel_prior if kernel_prior is not None else {}
+
+        self.jitter = Variable(self.X.data.new([jitter]))
 
     def model(self):
-        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, self.priors)
+        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, self.kernel_prior)
         kernel = kernel_fn()
-        K = kernel(self.X) + self.noise.repeat(self.input_dim).diag()
-        zero_loc = Variable(K.data.new([0]).expand(self.input_dim))
-        pyro.sample("f", dist.MultivariateNormal(zero_loc, K), obs=self.y)
+
+        K = kernel(self.X) + self.noise.expand(self.num_data).diag()
+        zero_loc = Variable(K.data.new([0])).expand(self.num_data)
+        pyro.sample("y", dist.MultivariateNormal(zero_loc, K), obs=self.y)
 
     def guide(self):
-        guide_priors = {}
-        for p in self.priors:
+        kernel_guide_prior = {}
+        for p in self.kernel_prior:
             p_MAP_name = pyro.param_with_module_name(self.kernel.name, p) + "_MAP"
             # init params by their prior means
-            p_MAP = pyro.param(p_MAP_name, Variable(self.priors[p].analytic_mean().data.clone(),
+            p_MAP = pyro.param(p_MAP_name, Variable(self.kernel_prior[p].torch_dist.mean.data.clone(),
                                                     requires_grad=True))
-            guide_priors[p] = dist.Delta(p_MAP)
-        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, guide_priors)
-        return kernel_fn()
+            kernel_guide_prior[p] = dist.Delta(p_MAP)
 
-    def forward(self, Z):
+        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, kernel_guide_prior)
+        kernel = kernel_fn()
+
+        return kernel
+
+    def forward(self, Xnew, full_cov=False, noiseless=False):
         """
-        Compute the parameters of `p(y|Z) ~ N(loc, covariance_matrix)`
-            w.r.t. the new input Z.
+        Compute the parameters of `p(y|Xnew) ~ N(loc, cov)`
+            w.r.t. the new input Xnew.
 
-        :param torch.autograd.Variable Z: A 2D tensor.
-        :return: loc and covariance matrix of p(y|Z).
+        :param torch.autograd.Variable Xnew: A 2D tensor.
+        :return: loc and covariance matrix of p(y|Xnew).
         :rtype: torch.autograd.Variable and torch.autograd.Variable
         """
-        if Z.dim() == 2 and self.X.size(1) != Z.size(1):
+        if Xnew.dim() == 2 and self.X.size(1) != Xnew.size(1):
             assert ValueError("Train data and test data should have the same feature sizes.")
-        if Z.dim() == 1:
-            Z = Z.unsqueeze(1)
+        if Xnew.dim() == 1:
+            Xnew = Xnew.unsqueeze(1)
+
         kernel = self.guide()
-        K = kernel(self.X) + self.noise.repeat(self.input_dim).diag()
+        Kff = kernel(self.X)
+        if noiseless:
+            Kff = Kff + self.jitter.expand(self.num_data).diag()
+        else:
+            Kff = Kff + self.noise.expand(self.num_data).diag()
```

Comment:
This is correct, however, it is safer to add jitter regardless, as sometimes the MAP solution may be attracted to short lengthscales and very low noise (aka Level-2 overfitting :))

Reply:
I think that it is better to address this when we start to put constraints on parameters (lengthscale, variance, noise,...).